examples/erc20-counter/apps/src/bin/publisher.rs

```diff
@@ -16,61 +16,69 @@
 // to the Bonsai proving service and publish the received proofs directly
 // to your deployed app contract.
 
-use std::time::Duration;
-
 use alloy::{
-    network::EthereumWallet, providers::ProviderBuilder, signers::local::PrivateKeySigner,
-    sol_types::SolCall,
+    network::EthereumWallet,
+    providers::ProviderBuilder,
+    signers::local::PrivateKeySigner,
+    sol_types::{SolCall, SolValue},
 };
 use alloy_primitives::Address;
 use anyhow::{ensure, Context, Result};
 use clap::Parser;
-use erc20_counter_methods::BALANCE_OF_ELF;
+use erc20_counter_methods::{BALANCE_OF_ELF, BALANCE_OF_ID};
 use risc0_ethereum_contracts::encode_seal;
 use risc0_steel::{
     ethereum::{EthEvmEnv, ETH_SEPOLIA_CHAIN_SPEC},
     host::BlockNumberOrTag,
-    Contract,
+    Contract, SolCommitment,
 };
-use risc0_zkvm::{default_prover, ExecutorEnv, ProverOpts, VerifierContext};
+use risc0_zkvm::{default_prover, sha::Digest, ExecutorEnv, ProverOpts, VerifierContext};
 use tokio::task;
 use tracing_subscriber::EnvFilter;
 use url::Url;
 
 alloy::sol! {
-    /// ERC-20 balance function signature.
-    /// This must match the signature in the guest.
+    /// Interface to be called by the guest.
     interface IERC20 {
         function balanceOf(address account) external view returns (uint);
     }
+
+    /// Data committed to by the guest.
+    struct Journal {
+        SolCommitment commitment;
+        address tokenContract;
+    }
 }
 
 alloy::sol!(
-    #[sol(rpc)]
-    "../contracts/ICounter.sol"
+    #[sol(rpc, all_derives)]
+    "../contracts/src/ICounter.sol"
 );
 
-/// Arguments of the publisher CLI.
+/// Simple program to create a proof to increment the Counter contract.
 #[derive(Parser, Debug)]
-#[clap(author, version, about, long_about = None)]
 struct Args {
-    /// Ethereum Node endpoint.
+    /// Private key
     #[clap(long, env)]
     eth_wallet_private_key: PrivateKeySigner,
 
-    /// Ethereum Node endpoint.
+    /// Ethereum RPC endpoint URL
     #[clap(long, env)]
-    rpc_url: Url,
+    eth_rpc_url: Url,
 
-    /// Counter's contract address on Ethereum
-    #[clap(long)]
-    contract: Address,
+    /// Beacon API endpoint URL
```

Comment:
```suggestion
    /// Optional Beacon API endpoint URL.
    ///
    /// If provided, the Steel proof will use a beacon block commitment, rather than the execution block.
    /// This allows proofs to be validated against a block hash from EIP-4788 with n expiration of about 1 day,
    /// rather than an execution block hash with an expiration of about 50 minutes.
```